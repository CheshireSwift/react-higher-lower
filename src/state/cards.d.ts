export type Suit = "H" | "S" | "C" | "D";
export type Rank = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;
export interface Card {
  rank: Rank;
  suit: Suit;
}
