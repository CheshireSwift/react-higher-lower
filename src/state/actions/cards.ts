import { Card } from "../cards";

export const draw = (card: Card) => ({
  type: "DRAW" as "DRAW",
  payload: card
});

export const shuffle = () => ({
  type: "SHUFFLE" as "SHUFFLE"
});
