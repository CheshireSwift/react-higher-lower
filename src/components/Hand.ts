import { div, h } from "react-hyperscript-helpers";
import { connect } from "react-redux";

import { State } from "../state";
import { Card } from "../state/cards";
import { CardDisplay } from "./CardDisplay";

interface HandProps {
  cards: Card[];
}

export const Hand = (props: HandProps) =>
  div(
    props.cards.map(card =>
      h(CardDisplay, {
        style: { display: "inline-block", marginRight: 8 },
        card
      })
    )
  );

export default connect((state: State) => ({ cards: state.hand }))(Hand);
