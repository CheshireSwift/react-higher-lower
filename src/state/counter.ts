import { increment, decrement, set } from "./actions/tick";

interface CounterState {
  count: number;
}

export const counter: Reducer<
  CounterState,
  typeof increment | typeof decrement | typeof set
> = (state = { count: 0 }, action) => {
  switch (action.type) {
    case "INCREMENT":
      return { ...state, count: state.count + action.payload.amount };
    case "DECREMENT":
      return { ...state, count: state.count - action.payload.amount };
    case "SET":
      return { ...state, count: action.payload.value };
    default:
      return state;
  }
};
