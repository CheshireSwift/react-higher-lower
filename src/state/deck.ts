import _ from "lodash";

import { Card, Rank, Suit } from "./cards";
import { draw, shuffle } from "./actions/cards";

type DeckState = Card[];

const sortedDeck = (): DeckState =>
  _.flatten(
    (["S", "H", "C", "D"] as Suit[]).map(suit =>
      ([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] as Rank[]).map(rank => ({
        suit,
        rank
      }))
    )
  );

export const deck: Reducer<DeckState, typeof draw | typeof shuffle> = (
  state = sortedDeck(),
  action
) => {
  switch (action.type) {
    case "DRAW":
      return _.reject(state, action.payload);
    case "SHUFFLE":
      return _.shuffle(state);
    default:
      return state;
  }
};
