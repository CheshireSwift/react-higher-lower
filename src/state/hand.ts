import _ from "lodash";

import { Card, Guess } from "./cards";
import { draw, guess } from "./actions/cards";

type HandState = Array<{ card: Card; guess: Guess | null }>;

export const hand: Reducer<HandState, typeof draw | typeof guess> = (
  state = [],
  action
) => {
  switch (action.type) {
    case "DRAW":
    case "GUESS":
      return [
        ...state,
        {
          card: action.payload.card,
          guess: _.get(action.payload, "direction", null)
        }
      ];
    default:
      return state;
  }
};
