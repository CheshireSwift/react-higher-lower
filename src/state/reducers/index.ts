import { combineReducers, Reducer } from "redux";

import { counter } from "./counter";

export type ReducerState<R> = R extends Reducer<infer S> ? S : null;
export const rootReducer = combineReducers({ counter });
