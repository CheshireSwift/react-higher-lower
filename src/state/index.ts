import { ReducerState, rootReducer } from "./reducers";
export type State = ReducerState<typeof rootReducer>;

export default State;
