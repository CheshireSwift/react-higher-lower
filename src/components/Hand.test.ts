import { render } from "enzyme";
import { h } from "react-hyperscript-helpers";

import { Card, Guess } from "../state/cards";

import { Hand } from "./Hand";

describe("the Hand", () => {
  const pairs: Array<{ card: Card; guess: Guess | null }> = [
    { card: { suit: "H", rank: 12 }, guess: null },
    { card: { suit: "C", rank: 4 }, guess: null },
    { card: { suit: "S", rank: 1 }, guess: null }
  ];
  let hand: Cheerio;

  beforeEach(() => {
    hand = render(h(Hand, { pairs }));
  });

  it("matches the snapshot", () => {
    expect(hand).toMatchSnapshot();
  });

  it("shows the cards in hand", () => {
    expect(hand.text()).toContain("QH");
    expect(hand.text()).toContain("4C");
    expect(hand.text()).toContain("AS");
  });
});
