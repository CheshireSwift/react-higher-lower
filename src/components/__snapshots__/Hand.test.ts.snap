// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Hand matches the snapshot 1`] = `
<div>
  <div
    style="display:inline-block;margin-right:8px;width:40px;height:60px;border:1px solid black;background:white;background-size:20px 20px"
  >
    QH
  </div>
  <div
    style="display:inline-block;margin-right:8px;width:40px;height:60px;border:1px solid black;background:white;background-size:20px 20px"
  >
    4C
  </div>
  <div
    style="display:inline-block;margin-right:8px;width:40px;height:60px;border:1px solid black;background:white;background-size:20px 20px"
  >
    AS
  </div>
</div>
`;
